programs/test/selftest.c

```diff
@@ -77,6 +79,103 @@
 #endif
 
 
+/* Sanity check for malloc. This is not expected to fail, and is rather
+ * intended to display potentially useful information about the platform,
+ * in particular the behavior of malloc(0). */
+static int calloc_self_test( int verbose )
+{
+    int failures = 0;
+    void *empty1 = mbedtls_calloc( 0, 1 );
+    void *empty2 = mbedtls_calloc( 0, 1 );
+    void *buffer1 = mbedtls_calloc( 1, 1 );
+    void *buffer2 = mbedtls_calloc( 1, 1 );
+    /* Attempt to allocate a size that's larger than size_t. This
+     * should fail. With AddressSanitizer, this crashes the program by
+     * default; to avoid this, set the environment variable
+     * ASAN_OPTIONS=allocator_may_return_null=1 when running this program.
+     */
+    void *overflow = mbedtls_calloc( SIZE_MAX / 2 + 1, 2 );
+    uintptr_t old_buffer1;
+
+    if( empty1 == NULL && empty2 == NULL )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(0): passed (NULL)\n" );
+    }
+    else if( empty1 == NULL || empty2 == NULL )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(0): failed (mix of NULL and non-NULL)\n" );
+        ++failures;
+    }
+    else if( empty1 == empty2 )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(0): passed (same non-null)\n" );
+    }
+    else
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(0): passed (distinct non-null)\n" );
+    }
+
+    if( buffer1 == NULL || buffer2 == NULL )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(1): failed (NULL)\n" );
+        ++failures;
+    }
+    else if( buffer1 == buffer2 )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(1): failed (same buffer twice)\n" );
+        ++failures;
+    }
+    else
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(1): passed\n" );
+    }
+
+    old_buffer1 = (uintptr_t) buffer1;
+    mbedtls_free( buffer1 );
+    buffer1 = mbedtls_calloc( 1, 1 );
+    if( buffer1 == NULL )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(1 again): failed (NULL)\n" );
+        ++failures;
+    }
+    else
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(1 again): passed (%s address)\n",
+                            (uintptr_t) old_buffer1 == (uintptr_t) buffer1 ?
+                            "same" : "different" );
+    }
+
+    if( overflow == NULL )
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(overflow): passed\n" );
+    }
+    else
+    {
+        if( verbose )
+            mbedtls_printf( "  CALLOC(overflow): failed\n" );
+        ++failures;
+    }
+
+    if( verbose )
+        mbedtls_printf( "\n" );
+    mbedtls_free( empty1 );
+    mbedtls_free( empty2 );
+    mbedtls_free( buffer1 );
+    mbedtls_free( buffer2 );
+    mbedtls_free( overflow );
+    return( failures );
```

Comment:
What sorts of platforms would this fail on? Why are we checking exactly? What should we or our users do if we find a platform where this self-test fails? (Implement a calloc wrapper to make it work?)

> The main objective is to report whether calloc returns NULL for a size
of 0.

Yes, but so what? What is to be done with this knowledge?

Reply:
It's a self-test. Self-tests don't check much that's useful. We don't expect the AES self-test to fail either — if it did, the unit tests would fail. The reason I added it here was to confirm that the test job that's supposed to test `calloc(0)=NULL` indeed has `calloc(0)=NULL`.

If the test fails, you know your `calloc` is broken. You know your platform better than we do, so we can't help you to fix it.

Reply:
What if we get `CALLOC(0): passed (distinct non-null)`? That is a valid and passing self-test result, isn't it? Does our other test fail to test what we want when that happens?

Reply:
`CALLOC(0): passed (distinct non-null)` is what you get on most if not all the platforms we test on today. I don't understand your question. The goal of this PR is to test on a platform with `CALLOC(0): passed (distinct non-null)`.

I didn't add a test run for `CALLOC(0): passed (same non-null)` because I don't think it matters to our code, whereas null vs non-null does.

Reply:
> The reason I added it here was to confirm that the test job that's supposed to test calloc(0)=NULL indeed has calloc(0)=NULL.

What if the platform doesn't have `calloc(0)=NULL`? Is the test job that's supposed to test `calloc(0)=NULL` invalid or shouldn't be run?

Reply:
I don't understand the question. The platform may or may not return NULL for malloc(0): both are valid. The objective of this PR is to make sure that we test the library on a platform where malloc(0) returns NULL. Since we only use platforms where malloc(0) is non-null, I simulate a platform where malloc(0) returns NULL by defining a wrapper function for `calloc` (we only use `calloc`, never `malloc`) that forcibly returns NULL when told to allocate 0 bytes. So now there is one component of `all.sh` that runs the unit tests with `calloc` returning NULL for 0 bytes. This component also runs the selftest and I confirm that I didn't mess up the way I set up the wrapper by visually inspecting that the selftest prints `NULL`. I didn't test that in `all.sh` because it's not worth the trouble to do it with our current tooling.

Reply:
> make sure that we test the library on a platform where malloc(0) returns NULL.

This is checked only manually? or is there a test that will fail if the platform's malloc(0) doesn't return NULL?

Reply:
Yes, I checked this manually. As a rule we don't have meta-tests that test that the tests are testing what we think they're testing, so I didn't do it here. It's like checking that a non-regression test fails once, and not keeping a CI run with a buggy product to check that it keeps catching the bug as the test code evolves.

Reply:
I wasn't requesting any meta tests. I was looking to ensure we weren't adding tests that had specific platform requirements and were running them on platforms that don't meet the requirements, whether self tests, test suites or otherwise.

I understand now what we are doing with the information from the self test, and that it is manual and not automatic. Thanks for explaining.